### PR TITLE
feat: implemented Info endpoint for spring-boot

### PR DESCRIPTION
**Pull request by prahaladhchandrahasan**
Trying to implement the Info endpoint for spring boot as requested in the issue https://github.com/hawtio/hawtio-integration/issues/99.

### CODE REVIEW

File: plugins/spring-boot/spring-boot.module.ts [hunk 1]
```diff
@@ -4,12 +4,14 @@
 /// <reference path="spring-boot.config.ts"/>
 /// <reference path="spring-boot.component.ts"/>
 /// <reference path="spring-boot.service.ts"/>
+/// <reference path="info/info.module.ts"/>
```
Comment (tadayosi) [inline plugins/spring-boot/spring-boot.module.ts:7]: lets care about the order of references. it should be put right after health.

File: plugins/spring-boot/info/info.service.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
+/// <reference path="info.ts"/>
+/// <reference path="../common/endpoint-mbean.ts"/>
+
+namespace SpringBoot {
+
+  export class InfoService {
+
+    constructor(private jolokiaService: JVM.JolokiaService, private humanizeService: Core.HumanizeService, private springBootService: SpringBootService) {
```
Comment (tadayosi) [inline plugins/spring-boot/info/info.service.ts:8]: we don't need `humanizeService` right now. lets drop it.

File: plugins/spring-boot/info/info.component.less [hunk 3]
```diff
@@ -0,0 +1,4 @@
+.spring-boot-info-main {
+  margin-bottom: 20px;
+
```
Comment (tadayosi) [inline plugins/spring-boot/info/info.component.less:3]: remove blank line

File: plugins/spring-boot/info/info.component.ts [hunk 4]
```diff
@@ -0,0 +1,67 @@
+/// <reference path="info.service.ts"/>
+/// <reference path="info.ts"/>
+
+namespace SpringBoot {
+
+  export class InfoController {
+    readonly reloadDelay = 20000;
+    reloadTask: ng.IPromise<any>;
+    info: Info;
+    info_pair: Info
```
Comment (tadayosi) [inline plugins/spring-boot/info/info.component.ts:10]: use camel case for variable names. also make fields private where possible.

File: plugins/spring-boot/info/info.component.ts [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+      const successfullyCancelled = this.$interval.cancel(this.reloadTask);
+      if (!successfullyCancelled) {
+        log.error('Failed to cancel Info data reload task');
+      }
+    }
+  }
+
+  export const infoComponent: angular.IComponentOptions = {
+    template: `
+      <div class="spring-boot-info-main">
```
Comment (tadayosi) [inline plugins/spring-boot/info/info.component.ts:48]: format the table. there is inconsistency in indentation.

File: plugins/spring-boot/info/info.component.ts [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+        <h1>Info</h1>
+        <table class="table table-striped table-bordered table-hover">
+            <thead>
+            <th> Attribute </th>
+            <th> Value </th>
+            </thead>
+          <tbody>
+            <tr ng-repeat = "item in $ctrl.info_pair">
```
Comment (tadayosi) [inline plugins/spring-boot/info/info.component.ts:56]: dont add spaces before and after `=`.

File: plugins/spring-boot/info/info.component.ts [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    loadDataPeriodically() {
+      this.loadData();
+      this.reloadTask = this.$interval(() => this.loadData(), this.reloadDelay);
+    }
+
+    loadData() {
+      this.infoService.getInfo()
+        .then(info => {
+          this.info = info
+          this.info_pair = _.toPairs(info)
```
Comment (tadayosi) [inline plugins/spring-boot/info/info.component.ts:33]: if we only use the `infoPair` in html we don't need `info` at all. I'd rename it as `info` and use it in html.

File: plugins/spring-boot/info/info.ts [hunk 8]
```diff
@@ -0,0 +1,6 @@
+namespace SpringBoot {
+
+  export class Info {
```
Comment (tadayosi) [inline plugins/spring-boot/info/info.ts:3]: lets do something here. that's why you needed `_.toPairs()` in controller. Instead, convert the raw data into this Info object in service and directly use the data structure in html.

File: plugins/spring-boot/info/info.service.ts [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+
+  export class InfoService {
+
+    constructor(private jolokiaService: JVM.JolokiaService, private springBootService: SpringBootService) {
+      'ngInject';
+    }
+
+    getInfo(): ng.IPromise<Info> {
+      log.debug('Fetch Info data');
+      const mbean: EndpointMBean = this.springBootService.getEndpointMBean(['infoEndpoint', 'Info'], ['info'])
```
Comment (tadayosi) [inline plugins/spring-boot/info/info.service.ts:14]: which one is really needed? `infoEndpoint` or `Info`?

**Review by tadayosi — CHANGES_REQUESTED** [review]

File: plugins/spring-boot/info/info.component.ts [hunk 10]
```diff
@@ -0,0 +1,76 @@
+/// <reference path="info.service.ts"/>
+/// <reference path="info.ts"/>
+
+namespace SpringBoot {
+  const flattenJSON = (obj = {}, res = {}, extraKey = '') => {
```
Comment (tadayosi) [inline plugins/spring-boot/info/info.component.ts:5]: lets move this to service

File: plugins/spring-boot/info/info.component.ts [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+    loadDataPeriodically() {
+      this.loadData();
+      this.reloadTask = this.$interval(() => this.loadData(), this.reloadDelay);
+    }
+
+    loadData() {
+      this.infoService.getInfo()
+        .then(info => {
+          this.info = info
+          this.info_pair = Object.entries(flattenJSON(JSON.parse(JSON.stringify(info))));
```
Comment (tadayosi) [inline plugins/spring-boot/info/info.component.ts:42]: lets move this to service.

File: plugins/spring-boot/info/info.service.ts [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+      return this.jolokiaService.execute(mbean.objectName, mbean.operation)
```
Comment (tadayosi) [inline plugins/spring-boot/info/info.service.ts:15]: service should do the  conversion work here.

File: plugins/spring-boot/info/info.ts (hunk 8 above)
Comment (prahaladhchandrahasan) [reply plugins/spring-boot/info/info.ts:3]: did it in the new commit

Comment (tadayosi) [reply plugins/spring-boot/info/info.ts:3]: where is it?

File: plugins/spring-boot/spring-boot.service.spec.ts [hunk 13]
```diff
@@ -28,19 +41,20 @@ describe("SpringBootService", () => {
...
     it("should return two tabs", () => {
       // given
-      workspace.treeContainsDomainAndProperties.and.returnValues(true, false, false, false, true, true);
+      workspace.treeContainsDomainAndProperties.and.returnValues(true, false, true, false);
       // when
       const tabs = springBootService.getTabs();
       // then
       expect(tabs[0]).toEqual(healthTab);
-      expect(tabs[1]).toEqual(traceTab);
+      expect(tabs[1]).toEqual(loggersTab);
```
Comment (tadayosi) [inline plugins/spring-boot/spring-boot.service.spec.ts:52]: why did you change this?  it changes the meaning of the test slightly. can't you make it pass with `traceTab`?

File: plugins/spring-boot/spring-boot.service.spec.ts [hunk 14]
```diff
@@ -28,19 +41,20 @@ describe("SpringBootService", () => {
...
     });
 
     it("should return one tab", () => {
       // given
-      workspace.treeContainsDomainAndProperties.and.returnValues(false, false, false, false, true, true);
+      workspace.treeContainsDomainAndProperties.and.returnValues(false, false, false, false,false, false, true, true);
```
Comment (tadayosi) [inline plugins/spring-boot/spring-boot.service.spec.ts:57]: add space before `false` in `,false`

File: plugins/spring-boot/spring-boot.service.spec.ts [hunk 15]
```diff
@@ -28,19 +41,20 @@ describe("SpringBootService", () => {
       expect(tabs[2]).toEqual(traceTab);
     });
 
+
     it("should return two tabs", () => {
       // given
-      workspace.treeContainsDomainAndProperties.and.returnValues(true, false, false, false, true, true);
+      workspace.treeContainsDomainAndProperties.and.returnValues(true, false, true, false);
```
Comment (tadayosi) [inline plugins/spring-boot/spring-boot.service.spec.ts:47]: this change doesn't look correct.

File: plugins/spring-boot/spring-boot.service.spec.ts [hunk 16]
```diff
@@ -19,7 +20,19 @@ describe("SpringBootService", () => {
 
     it("should return all tabs", () => {
       // given
-      workspace.treeContainsDomainAndProperties.and.returnValues(true, false, true, false, true);
+      workspace.treeContainsDomainAndProperties.and.returnValues(true,true, true, true);
```
Comment (tadayosi) [inline plugins/spring-boot/spring-boot.service.spec.ts:23]: this doesn't look correct.

**Review by tadayosi — CHANGES_REQUESTED** [review]
